### PR TITLE
🧑‍💻 Add Notebook URL Replacement Command

### PR DESCRIPTION
**Pull request by John-P**
Adds a script to update notebook URLs to match the current git ref.

This can be run with `make`. This will pass in all files in `examples`, uses the current branch name as the 'to' ref, and "develop" as the 'from' ref.
```bash
make update-notebook-urls
```

You can also run the script manually. If run manually, you can specify the target ref and input files.

```bash
python pre-commit/notebook_urls.py --from develop --to dev-notebook-urls-hook examples/*
```

### CODE REVIEW

File: examples/01-wsi-reading.ipynb [hunk 1]
```diff
@@ -73,7 +73,7 @@
    "outputs": [],
```
Comment (shaneahmed) [inline examples/01-wsi-reading.ipynb:73]: <p>Why it is replaced with @HEAD?</p><p>This doesn't seem right.</p><p>Also there are more changes in the rest of the notebooks. Any reasons why?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TissueImageAnalytics/tiatoolbox/pull/519/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (shaneahmed) [reply examples/01-wsi-reading.ipynb:73]: I have tried different options but it didn't work.
Can we not pass it as an argument?

File: pre-commit/notebook_check_ast.py [hunk 2]
```diff
@@ -16,7 +16,7 @@ def main(files: List[Path]) -> bool:
         for n, cell in enumerate(notebook["cells"]):
             if cell["cell_type"] != "code":
                 continue
-            source = "".join([x for x in cell["source"] if not x.startswith("!")])
+            source = "".join([x for x in cell["source"] if x[0] not in ("#", "%", "!")])
```
Comment (shaneahmed) [inline pre-commit/notebook_check_ast.py:19]: This is causing it to fail on Windows.

Comment (John-P) [reply pre-commit/notebook_check_ast.py:19]: Can you give any more detail? Like the error output? I cannot see any reason why this should behave any differently on a different platform. I need more information.

Comment (shaneahmed) [reply pre-commit/notebook_check_ast.py]: `
check notebook ast.......................................................Failed
- hook id: notebook-check-ast
- exit code: 1
Traceback (most recent call last):
  File "pre-commit/notebook_check_ast.py", line 33, in <module>
    sys.exit(1 - main(args.files))
  File "pre-commit/notebook_check_ast.py", line 15, in main
    notebook = json.load(fh)
  File "c:\tools\anaconda3\envs\tiatoolbox_dev\lib\json\__init__.py", line 293, in load
    return loads(fp.read(),
  File "c:\tools\anaconda3\envs\tiatoolbox_dev\lib\encodings\cp1252.py", line 23, in decode
    return codecs.charmap_decode(input,self.errors,decoding_table)[0]
UnicodeDecodeError: 'charmap' codec can't decode byte 0x9d in position 4641: character maps to <undefined>
notebook urls............................................................Failed
- hook id: notebook-urls
- exit code: 1
examples\01-wsi-reading.ipynb
Traceback (most recent call last):
  File "pre-commit/notebook_urls.py", line 93, in <module>
    sys.exit(1 - main(args.files))
  File "pre-commit/notebook_urls.py", line 55, in main
    notebook = json.load(fh)
  File "c:\tools\anaconda3\envs\tiatoolbox_dev\lib\json\__init__.py", line 293, in load
    return loads(fp.read(),
  File "c:\tools\anaconda3\envs\tiatoolbox_dev\lib\encodings\cp1252.py", line 23, in decode
    return codecs.charmap_decode(input,self.errors,decoding_table)[0]
UnicodeDecodeError: 'charmap' codec can't decode byte 0x9d in position 4641: character maps to <undefined>
notebook markdown format.................................................Failed
- hook id: notebook-markdown-format
- exit code: 1
Traceback (most recent call last):
  File "pre-commit/notebook_markdown_format.py", line 74, in <module>
    main(sorted(args.files))
  File "pre-commit/notebook_markdown_format.py", line 52, in main
    notebook = json.loads(path.read_text())
  File "c:\tools\anaconda3\envs\tiatoolbox_dev\lib\pathlib.py", line 1222, in read_text
    return f.read()
  File "c:\tools\anaconda3\envs\tiatoolbox_dev\lib\encodings\cp1252.py", line 23, in decode
    return codecs.charmap_decode(input,self.errors,decoding_table)[0]
UnicodeDecodeError: 'charmap' codec can't decode byte 0x9d in position 4641: character maps to <undefined>
`

File: examples/01-wsi-reading.ipynb (hunk 1 above)
Comment (John-P) [reply examples/01-wsi-reading.ipynb:73]: <p>HEAD is the current ref (e.g. the branch name) see, <a href="https://initialcommit.com/blog/what-is-git-head#git-refs-and-heads" target="_blank">https://initialcommit.com/blog/what-is-git-head#git-refs-and-heads</a></p>

Comment (John-P) [reply examples/01-wsi-reading.ipynb:73]: <p>Pass what as an argument?</p>

File: pre-commit/notebook_check_ast.py (hunk 2 above)
Comment (John-P) [reply pre-commit/notebook_check_ast.py]: Ok thank you, this traceback indicates the issue is not on this line. It is a file encoding issue. It appears windows is not using utf-8 for some reason.

Comment (John-P) [reply pre-commit/notebook_check_ast.py]: See https://github.com/TissueImageAnalytics/tiatoolbox/pull/519/commits/f9de1fa39a457921343cc18ae22300ccdb5e6c92

File: examples/01-wsi-reading.ipynb (hunk 1 above)
Comment (shaneahmed) [reply examples/01-wsi-reading.ipynb:73]: Pass branch name as an argument, if possible.

Comment (shaneahmed) [reply examples/01-wsi-reading.ipynb:73]: > HEAD is the current ref (e.g. the branch name) see, https://initialcommit.com/blog/what-is-git-head#git-refs-and-heads. It appears that the command to get the current ref is returning the shorthand HEAD instead of the branch name. I am not able to replicate this. Can you provide steps?

This happens online only. Locally it runs fine.

File: Makefile [hunk 3]
```diff
@@ -84,3 +84,6 @@ dist: clean ## builds source and wheel package
 
 install: clean ## install the package to the active Python's site-packages
 	python setup.py install
+
+update-notebook-urls:
+	python pre-commit/notebook_urls.py examples/*
```
Comment (shaneahmed) [inline Makefile:89]: I have changed `examples/**.ipynb` to `examples/*` as `rglob` was not reading files in subfolders properly.
I have also updated the description.

File: pre-commit/notebook_urls.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+        from_ref:
+            Reference to diff from.
+        to_ref:
+            Reference to diff to.
+
+    """
+    replacements = [
+        PatternReplacement(
+            pattern=(
+                r"(^\s*[!%]\s*)pip install "
```
Comment (shaneahmed) [inline pre-commit/notebook_urls.py:90]: This only replaces `pip install ...` commands. Please can you also add replacements for links to GitHub and Colab at the start of each notebook. E.g., `https://github.com/TissueImageAnalytics/tiatoolbox/blob/develop/examples/01-wsi-reading.ipynb` should be replaced by `https://github.com/TissueImageAnalytics/tiatoolbox/blob/[dev-notebook-urls-hook](https://github.com/TissueImageAnalytics/tiatoolbox/tree/dev-notebook-urls-hook)/examples/01-wsi-reading.ipynb`. Same for Colab links.

**Review by shaneahmed — CHANGES_REQUESTED** [review]
Thanks John. Looks great. Please can you add replacements for GitHub and Colab links as well?
Also the json.load fails on 10-wsi-registration notebook. Please can you update this?

File: pre-commit/notebook_urls.py [hunk 5]
```diff
@@ -110,6 +110,10 @@ def main(files: List[Path], from_ref: str, to_ref: str) -> bool:
     passed = True
     print(f"From ref '{from_ref}' to ref '{to_ref}'")
     for path in files:
+        if path.suffix != ".ipynb":
+            print(f"Skipping {path} (not a Jupyter Notebook).")
+            return passed
+
```
Comment (John-P) [inline pre-commit/notebook_urls.py:116]: The file extension does not make it a jupyter notebook. We should not rely on this.

File: Makefile [hunk 6]
```diff
@@ -86,4 +86,6 @@ install: clean ## install the package to the active Python's site-packages
 	python setup.py install
 
 update-notebook-urls:
-	python pre-commit/notebook_urls.py examples/*
+	python pre-commit/notebook_urls.py examples/*.ipynb
+	python pre-commit/notebook_urls.py examples/inference-pipelines/*.ipynb
+	python pre-commit/notebook_urls.py examples/full-pipelines/*.ipynb
```
Comment (John-P) [inline Makefile]: A generic glob pattern would probably be better, this requires manually updating whenever a new subdirectory is added.

File: Makefile (hunk 3 above)
Comment (John-P) [reply Makefile]: I've updated this to use find and xargs to recursively find .ipynb files in examples so that we do not have to manually add every subdirectory.

File: pre-commit/notebook_urls.py [hunk 7]
```diff
@@ -204,4 +209,4 @@ def replace_line(line: str, to_ref: str, replacements: List[PatternReplacement])
         default=git_branch_name(),
     )
     args = parser.parse_args()
-    sys.exit(1 - main(args.files, args.from_ref, args.to_ref))
+    sys.exit(main(args.files, args.from_ref, args.to_ref))
```
Comment (John-P) [inline pre-commit/notebook_urls.py]: Why is this changed? It will cause the program to exit with code 1 when passing. It should be 0 on success.

Comment (John-P) [reply pre-commit/notebook_urls.py]: Generating an error where? It's supposed to be 0 on success

Comment (shaneahmed) [reply pre-commit/notebook_urls.py]: OK, now it runs fine.

**Comment by shaneahmed** [comment]
The rest of the PR is fine except the URL replacement which is not working still.

**Comment by shaneahmed** [comment]
> The rest of the PR is fine except the URL replacement which is not working still.

@John-P I think this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/519/commits/810883f5b1ee99e4d724f0d29192bd21ec023570 would fix the issue with URL. Please can you double check before I merge this?

**Review by shaneahmed — APPROVED** [review]
Thanks @John-P
